query/expression.go

```diff
@@ -144,19 +144,66 @@ func (expr *metricFetchExpression) Evaluate(context EvaluationContext) (value, e
 }
 
 func (expr *functionExpression) Evaluate(context EvaluationContext) (value, error) {
-	switch expr.functionName {
+
+	name := expr.functionName
+	switch name {
 	case "+":
-		return evaluateBinaryOperation(context, expr.functionName, expr.arguments,
-			func(left, right float64) float64 { return left + right })
+		fallthrough
 	case "-":
-		return evaluateBinaryOperation(context, expr.functionName, expr.arguments,
-			func(left, right float64) float64 { return left - right })
-	case "/":
-		return evaluateBinaryOperation(context, expr.functionName, expr.arguments,
-			func(left, right float64) float64 { return left / right })
+		fallthrough
 	case "*":
-		return evaluateBinaryOperation(context, expr.functionName, expr.arguments,
-			func(left, right float64) float64 { return left * right })
+		fallthrough
+	case "/":
+		arguments := make([]value, len(expr.arguments))
```

Comment:
Although it's possible to move this definition up to the top of the structure (especially since it's repeated for the aggregate functions), in order to change the `Timerange` to accommodate time shifting, arguments can't be evaluated until the function name is known (and then they may need to be evaluated in a special way).